### ISSUE TITLE
Reduce compile times

### ISSUE THREAD

**Issue by ndmitchell** [digital-asset/ghc-lib#39]
For projects that only need the parser, the compile time is a bit ferocious. Would be interesting if we can create ghc-lib-parser which is just enough to get as far as the compiler, and see if that reduces the compile times by pruning the module graph.

**Comment by neil-da** [comment]
Suggestion from @shayne-fletcher-da was that you could have `ghc-lib-parser` with only things required for the parser, and then have `ghc-lib` depend on that.

**Comment by ndmitchell** [comment]
From the GSoC Hadrian speed project, it was identified that DynFlags and HsInstances both go really slow because of optimisations. If you add `{-# OPTIONS_GHC -O0 #-}` to the top of the file you save about 5 minutes from compilation, and it doesn't have that measurable an impact on the resulting program. I suggest that ghc-lib generate those headings. CC @shayne-fletcher-da

**Comment by shayne-fletcher-da** [comment]
Nice. That is a massive saving in compile times for `ghc-lib-parser`. Let's do it! @ndmitchell  I will take your suggestion and produce a PR.

**Comment by neil-da** [comment]
I would also suggest benchmarking before and after, to confirm it really makes a difference in practice.

**Comment by shayne-fletcher-da** [comment]
TL;DR this saves us a few minutes so appears to be worth it. PR implementing selective optimization disabling [here](https://github.com/digital-asset/ghc-lib/pull/90).
```
As is (disable optimizations for none):
ghc-lib-parser  7m41.556s    7m51.356s
ghc-lib         8m4.639s     8m8.880s

Selective (disable optimizations just for `DynFlags.hs` and `HsIntances.hs`):
ghc-lib-parser  5m54.290s    5m44.188s
ghc-lib         7m13.772s    7m11.231s

Global (disable optimizations for all):
ghc-lib-parser  2m27.748s    2m26.850s
ghc-lib         2m51.471s    2m56.943s
```

**Comment by ndmitchell** [comment]
Now I wonder about disabling optimisations globally. I'd be curious the impact on HLint and on compiling a big DAML model.

**Comment by ndmitchell** [comment]
I should definitely disable optimisation on the HLint CI...

**Comment by shayne-fletcher-da** [comment]
```
Comparison of branches:
  - master
  - ghc-lib-8.8.0.20190704-selective-optimizations-off
  - ghc-lib-8.8.0.20190704.1-all-optimizations-off
Build times verified consistent with expectations (e.g. the last one is fast).
Results for
 `time bazel run //daml-foundations/daml-ghc:daml-ghc-test-dev -- --pattern=`
(last col. is mean):

8.8.0.20190612                             | 31.710s 30.102s 30.021s 29.616s | 30.362s
8.8.0.20190704-selective-optimizations-off | 30.876s 32.406s 32.252s 31.486s | 31.755s
8.8.0.20190704.1-all-optimizations-off     | 36.939s 35.454s 36.021s 35.269s | 35.921s
```

**Comment by shayne-fletcher-da** [comment]
Turning off optimizations basically doubles the execution time.
```
Results for Spider `time daml build`:
8.8.0.20190612                             | 3m25.391s 3m25.164s | 3m25.28
8.8.0.20190704-selective-optimizations-off | 3m17.858s 3m17.276s | 3m17.56
8.8.0.20190704.1-all-optimizations-off     | 5m48.609s 5m47.950s | 5m47.30
```

**Comment by shayne-fletcher-da** [comment]
```
* 8.8.0.20190612

LSP
  diagnostics
    diagnostics disappear after error is fixed:    OK (1.22s)
    diagnostics appear after introducing an error: OK (1.07s)
    percent encoding does not matter:              OK (1.06s)
    failed name resolution:                        OK (1.08s)
    import cycle:                                  OK (0.39s)
    import error:                                  OK (0.39s)
    multi module funny:                            OK (1.09s)
    parse error:                                   OK (1.09s)
    type error:                                    OK (1.10s)
    scenario error:                                OK (3.60s)
  requests
    code-lenses:                                   OK (1.26s)
    type on hover: name:                           OK (0.43s)
    type on hover: literal:                        OK (0.41s)
    definition:                                    OK (0.44s)
  scenarios
    opening codelens produces a notification:      OK (3.12s)
    scenario ok:                                   OK (3.13s)
    spaces in path:                                OK (3.48s)
  Stress tests
    Modify a file 2000 times:                      OK (5.08s)
    Set 10 files of interest:                      OK (1.34s)
    Type check 100-deep module chain:              OK (1.78s)

All 20 tests passed (32.55s)

* 8.8.0.20190704-selective-optimizations-off

LSP
  diagnostics
    diagnostics disappear after error is fixed:    OK (1.73s)
    diagnostics appear after introducing an error: OK (1.19s)
    percent encoding does not matter:              OK (1.17s)
    failed name resolution:                        OK (1.18s)
    import cycle:                                  OK (0.43s)
    import error:                                  OK (0.42s)
    multi module funny:                            OK (1.15s)
    parse error:                                   OK (1.15s)
    type error:                                    OK (1.38s)
    scenario error:                                OK (3.48s)
  requests
    code-lenses:                                   OK (1.31s)
    type on hover: name:                           OK (0.48s)
    type on hover: literal:                        OK (0.45s)
    definition:                                    OK (0.56s)
  scenarios
    opening codelens produces a notification:      OK (2.29s)
    scenario ok:                                   OK (3.44s)
    spaces in path:                                OK (3.35s)
  Stress tests
    Modify a file 2000 times:                      OK (4.93s)
    Set 10 files of interest:                      OK (1.37s)
    Type check 100-deep module chain:              OK (1.85s)

All 20 tests passed (33.29s)

* 8.8.0.20190704.1-all-optimizations-off

LSP
  diagnostics
    diagnostics disappear after error is fixed:    OK (1.31s)
    diagnostics appear after introducing an error: OK (1.18s)
    percent encoding does not matter:              OK (1.17s)
    failed name resolution:                        OK (1.15s)
    import cycle:                                  OK (0.41s)
    import error:                                  OK (0.42s)
    multi module funny:                            OK (1.19s)
    parse error:                                   OK (1.16s)
    type error:                                    OK (1.19s)
    scenario error:                                OK (3.67s)
  requests
    code-lenses:                                   OK (1.31s)
    type on hover: name:                           OK (0.65s)
    type on hover: literal:                        OK (0.50s)
    definition:                                    OK (0.69s)
  scenarios
    opening codelens produces a notification:      OK (2.18s)
    scenario ok:                                   OK (3.21s)
    spaces in path:                                OK (3.34s)
  Stress tests
    Modify a file 2000 times:                      OK (7.11s)
    Set 10 files of interest:                      OK (1.50s)
    Type check 100-deep module chain:              OK (2.28s)

All 20 tests passed (35.61s)
```

**Comment by shayne-fletcher-da** [comment]
Selective dis-optimizations landed in https://github.com/digital-asset/ghc-lib/pull/90.

**Comment by shayne-fletcher-da** [comment]
> I should definitely disable optimisation on the HLint CI...

Most definitely. How would we go about doing that? (For the record, for mucking around in `ghc-lib` itself, the easiest way I know is `cabal install ghc-lib-parser --disable-optimization`.)

**Comment by jinwoo** [comment]
`ghc-lib-parser` indeed takes long to build. It often times out in our build environment with optimization on. We reduced the optimization level down to `-O1` from `-O2` but it still sometimes times out. We recently upgraded to 8.10 and it may have made the build time worse. So we may have to reduce the level further to `-O0`.

Do you know how much the run-time speed of `ghc-lib-parser` is affected by the optimization level? I want to know how much `hlint` would be affected if we have to do that.

Thanks!

**Comment by shayne-fletcher** [comment]
If you are on 8.10 then you unless you manipulate flags, you don't need to build ghc-lib-parser at all since hlint (and ghc-lib-parser-ex) will bind to the native ghc libs in that case.

If you do have to use ghc-lib-parser, then there is a subset of files in the package that we disable optimization for at the source level. These particular files take a long time to compile and disabling optimization does not have a significant effect on runtime performance. They were identified in a Google summer of code project.

**Comment by jinwoo** [comment]
I'm sorry I meant 8.10 of `ghc-lib-parser`, not of GHC :) We're still using GHC 8.6.4 :(

Okay. So if optimization is disabled at source level, then even if we reduce the optimization level further via the command-line flag, it may not reduce the build time much (as well as the run-time).

We'll see. Thanks for the quick reply!

**Comment by shayne-fletcher** [comment]
Exactly. I expect if you disable optimization globally you'll not see much difference in build times but risk adversely affecting execution HLint execution times. Do let us know though!

**Comment by shayne-fletcher** [comment]
@garyverhaegen-da this is no longer active. close please
